Fix bug in SP* transactions

janardhan1993 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/2220)
<!-- Reviewable:end -->

janardhan1993 commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[contrib/integration/upsertDelete/main.go, line 1 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220#-L7Uy4H0wB7GwULmqLT-:-L7Uy4H0wB7GwULmqLT0:b-bemic4) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/contrib/integration/upsertDelete/main.go#L1)):*
> ```Go
> package main
> ```

Tested manually with this program, implementation of jepsen delete test.  Will refactor a bit and add to integration test later. Don't see this file.

---

*[posting/list.go, line 438 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220#-L7UyJ92w1HnbzglLbVZ:-L7UyJ92w1HnbzglLbV_:b-wk2jko) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L438)):*
> ```Go
> 	}
> 	// Check if this commit is for sp*, markdeleteAll stores the startTs for sp*
> 	if l.markdeleteAll == startTs {
> ```

Say there are two parallel transactions one sp* and other which adds uid both did mutate, then if we were doing sp* irrespective of which transaction was committed.

---

*[posting/list.go, line 540 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220#-L7UyfJthUQMOTo7JfCm:-L7UyfJthUQMOTo7JfCn:b-8cscoj) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L540)):*
> ```Go
> 		// Fixing the pl is difficult with locks.
> 		// Ignore if SP* was committed with timestamp >= readTs
> 		if ts := Oracle().CommitTs(l.markdeleteAll); ts < readTs {
> ```

Need to ignore sp* if it was committed after the read transaction had started.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[contrib/integration/upsertDelete/main.go, line 1 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220#-L7Uy4H0wB7GwULmqLT-:-L7WWM9YFSuqRfTbPds3:b-c5bo5f) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/contrib/integration/upsertDelete/main.go#L1)):*
<details><summary><i>Previously, janardhan1993 (Janardhan Reddy) wrote…</i></summary><blockquote>

Tested manually with this program, implementation of jepsen delete test.  Will refactor a bit and add to integration test later. Don't see this file.
</blockquote></details>

Add a TODO to make this as a test and not a main program.

---

*[posting/list.go, line 438 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220#-L7UyJ92w1HnbzglLbVZ:-L7WTc6YiK8yK1X-eorh:b-p7n7pn) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L438)):*
<details><summary><i>Previously, janardhan1993 (Janardhan Reddy) wrote…</i></summary><blockquote>

Say there are two parallel transactions one sp* and other which adds uid both did mutate, then if we were doing sp* irrespective of which transaction was committed.
</blockquote></details>

Yeah, I saw this earlier and thought it to be a bug but didn't get around to talking to you about it.

---

*[posting/list.go, line 535 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220#-L7WUZA881inz0Qgc2EU:-L7WUZA881inz0Qgc2EV:bm4woqd) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L535)):*
> ```Go
> 	} else if l.markdeleteAll == readTs {
> 		// Check if there is uncommitted sp* at current readTs.
> 		deleteTs = readTs
> ```

What is strange is that `deleteTs` is a `readTs` here and a `commitTs` below. It is used to compare against a `commitTs` in `inSnapshot`.

---

*[posting/list.go, line 538 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220#-L7WVpYCj3SMMBo9d5VP:-L7WVpYCj3SMMBo9d5VQ:b-t951nj) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L538)):*
> ```Go
> 	} else if l.markdeleteAll < readTs {
> 		// Ignore all reads before this.
> 		// Fixing the pl is difficult with locks.
> ```

Is this comment still valid? I don't understand what it means.

---

*[posting/list.go, line 539 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220#-L7WViKbhRGnKIqXPCxX:-L7WViKbhRGnKIqXPCxY:bn3pi6s) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L539)):*
> ```Go
> 		// Ignore all reads before this.
> 		// Fixing the pl is difficult with locks.
> 		// Ignore if SP* was committed with timestamp >= readTs
> ```

I don't think anything can get committed with timestamp = readTs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/2220)*
<!-- Sent from Reviewable.io -->

janardhan1993 commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[posting/list.go, line 438 at r1](https://beta.reviewable.io/reviews/dgraph-io/dgraph/2220#-L7UyJ92w1HnbzglLbVZ:-L80rE3R2AkWfkvfwCrA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L438)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Yeah, I saw this earlier and thought it to be a bug but didn't get around to talking to you about it.
</blockquote></details>

Done.

---

*[posting/list.go, line 535 at r1](https://beta.reviewable.io/reviews/dgraph-io/dgraph/2220#-L7WUZA881inz0Qgc2EU:-L81D89Q1i2Cl69qUnVR:b-22vq8p) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L535)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

What is strange is that `deleteTs` is a `readTs` here and a `commitTs` below. It is used to compare against a `commitTs` in `inSnapshot`.
</blockquote></details>

We should consider sp* if read is happening as part of same transaction which did sp*, in this case we would consider everything below that as deleted.
We should only consider sp* only if it was commited before readTs

---

*[posting/list.go, line 538 at r1](https://beta.reviewable.io/reviews/dgraph-io/dgraph/2220#-L7WVpYCj3SMMBo9d5VP:-L81DNNpEDoCaLxDGnut:b5y8u68) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L538)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Is this comment still valid? I don't understand what it means.
</blockquote></details>

In other cases we updated the committs in posting using atomic variable, but we can't do that in case of sp* because sometimes we have read lock we can't fix the posting(delelte everything and make it empty)

---

*[posting/list.go, line 539 at r1](https://beta.reviewable.io/reviews/dgraph-io/dgraph/2220#-L7WViKbhRGnKIqXPCxX:-L7ZYIr92_NHNtbzJnMi:bc3bl1m) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/posting/list.go#L539)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

I don't think anything can get committed with timestamp = readTs.
</blockquote></details>

Yeah, will update comment

---

*[contrib/integration/upsertDelete/main.go, line 1 at r1](https://beta.reviewable.io/reviews/dgraph-io/dgraph/2220#-L7Uy4H0wB7GwULmqLT-:-L80rAWL1xnLL9XddpzT:buy94bo) ([raw file](https://github.com/dgraph-io/dgraph/blob/70e712b251ec9ac1da220a6eed376aa15f11dce4/contrib/integration/upsertDelete/main.go#L1)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Add a TODO to make this as a test and not a main program.
</blockquote></details>

Converted to Test

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/dgraph-io/dgraph/2220)*
<!-- Sent from Reviewable.io -->
